src/testoptix/testoptix.cpp

```diff
@@ -0,0 +1,397 @@
+                NULL);
+    if (ap.parse(argc, argv) < 0) {
+        std::cerr << ap.geterror() << std::endl;
+        ap.usage ();
+        exit (EXIT_FAILURE);
+    }
+    if (help) {
+        std::cout <<
+            "testoptix -- Simple test for OptiX functionality\n"
+             OSL_COPYRIGHT_STRING "\n";
+        ap.usage ();
+        exit (EXIT_SUCCESS);
+    }
+    if (imagefile.empty()) {
+        std::cerr << "testrender: Must specify a filename for output render\n";
+        ap.usage();
+        exit (EXIT_FAILURE);
+    }
+}
+
+
+// Helper function to convert from Imath::Vec3 to optix::float3
+optix::float3 vec3_to_float3 (const Vec3& vec)
+{
+    return optix::make_float3 (vec.x, vec.y, vec.z);
+}
+
+
+// Copies the specified device buffer into an output vector, assuming that
+// the buffer is in FLOAT3 format (and that Vec3 and float3 have the same
+// underlying representation).
+std::vector<OSL::Color3>
+get_pixel_buffer (const std::string& buffer_name, int width, int height)
+{
+    const OSL::Color3* buffer_ptr =
+        static_cast<OSL::Color3*>(optix_ctx[buffer_name]->getBuffer()->map());
+
+    if (! buffer_ptr) {
+        std::cerr << "Unable to map buffer " << buffer_name << std::endl;
+        exit (EXIT_FAILURE);
+    }
+
+    std::vector<OSL::Color3> pixels;
+    std::copy (&buffer_ptr[0], &buffer_ptr[width * height], back_inserter(pixels));
+
+    optix_ctx[buffer_name]->getBuffer()->unmap();
+
+    return pixels;
+}
+
+
+// Compiles a CUDA source file to PTX using NVRTC and returns it as a string.
+void get_ptx_from_cu_file (std::string& ptx_string, const char* filename)
+{
+    // Read the CUDA source file into a std::string
+    std::string cu_source;
+    if (! OIIO::Filesystem::read_text_file (filename, cu_source)) {
+        std::cerr << "Unable to load " << filename << std::endl;
+        exit (EXIT_FAILURE);
+    }
+
+    // Set up the default NVRTC options
+    std::vector<const char *> options;
+    const std::string src_dir    = std::string("-I") + OIIO::Filesystem::parent_path (filename);
+    const std::string cuda_inc   = std::string("-I") + CUDA_INCLUDE_DIR;
+    const std::string optix_inc  = std::string("-I") + OPTIX_INCLUDE_DIR;
+    const std::string optixu_inc = std::string("-I") + OPTIX_INCLUDE_DIR + "/optixu";
+
+    options.push_back (src_dir.c_str());
+    options.push_back (cuda_inc.c_str());
+    options.push_back (optix_inc.c_str());
+    options.push_back (optixu_inc.c_str());
+
+    // NB: The 'gpu-architecture' option specifies the minimum compute
+    //     capability required to execute the generated PTX and may need
+    //     to be changed depending on the requirements of the input CUDA.
+    options.push_back ("--gpu-architecture=compute_35");
+    options.push_back ("--use_fast_math");
+    options.push_back ("--device-as-default-execution-space");
+    options.push_back ("--relocatable-device-code=true");
+    options.push_back ("-D__x86_64");
```

Comment:
This seems to also need "-D__x86_64__"

Reply:
An alternative to using NVRTC would be using NVCC to compile the CUDA sources to PTX as part of the build process, and have testoptix load the PTX files at run time. It would remove some complexity from testoptix and avoid recompiling the CUDA sources each run (although the overhead is small for these simple cases). I can implement either approach.

Reply:
Well, remember that part of the purpose of testoptix is (ideally) to have a clear example that people can look over to figure out how to integrate OptiX+OSL into their own renderer. So I would advise choosing based on what you want to promote as best practices.